utils/hooks: Fix import error with GTK apps

danyeaw opened the pull request:
Fixes #6298 by adding imports that were missing from the isolated environment. Add Ubuntu dependencies so that the
test_gi tests will be run during CI.

danyeaw committed:
utils/hooks: Fix import error with GTK apps.

Fixes #6298 by adding imports that were missing from the
isolated environment. Add Ubuntu dependencies so that the
test_gi tests will be run during CI.

bwoodsend commented on news/6300.bugfix.rst:
```diff
@@ -0,0 +1,2 @@
+Fix import errors when calling ``get_gi_libdir()`` during packaging of GTK apps.
+Enable CI tests of GTK by adding PyGObject dependencies for the Ubuntu builds.
```
The broken behaviour was never released so technically we don't need a news entry at all. That said, so many people install directly from the `develop` branch that every commit to `develop` is a release in a sense so I think we should have the news entry anyway?

danyeaw replied:
Yup, I was considering if it made it to develop then it is now a bug fixed. I am fine with removing the news entry as well if we don't normally do that.

rokm replied:
We can keep the news fragment in, especially now that we have real-time preview of [future release changelog](https://pyinstaller.readthedocs.io/en/latest/CHANGES.html).

But, I would keep the second line out of the bugfix news fragment. 

And perhaps change the first one to something along the lines of "Fix regression in ``gi``  hooks introduced by transition to the new process-isolation framework.``  to make it clear that this is a "new" problem and not something that has been around for a while.

rokm commented on .github/workflows/ci.yml:
```diff
@@ -46,7 +46,9 @@ jobs:
             libopengl0 libegl1 \
-            libpulse0 libpulse-mainloop-glib0
+            libpulse0 libpulse-mainloop-glib0 \
+            python3-dev python3-gi python3-gi-cairo gir1.2-gtk-3.0 \
```
`python3-*` packages are redundant here, because CI is not running the system version of python. So we need to install `gtk3` packages and their `dev` counterparts via `apt`, and install/build `python-gobject` into CI's python environment via `pip`.

danyeaw replied:
@rokm Thanks, this made me realize I can remove some dependencies for all the PyGObject projects I work on 👍. I think I now added the bare minimum and I added PyGObject to the `requirements-libraries.txt` with a `sys_platform` constraint.

rokm commented:
Looks like the `gi` tests are indeed enabled now :)

rokm commented:
We have another bug here:

https://github.com/pyinstaller/pyinstaller/blob/7234ba0e8b087cdee7f1ca0a3fe05c5967e738c3/PyInstaller/utils/hooks/gi.py#L31-L36

`repo.get_shared_library(module)` returns [a single name](https://lazka.github.io/pgi-docs/#GIRepository-2.0/classes/Repository.html#GIRepository.Repository.get_shared_library), so we must not iterate over it, but rather check for `None`.

The iteration comes from direct port of `v4`:
https://github.com/pyinstaller/pyinstaller/blob/1f9a2890dfa99525549bbbfdb985352154bff492/PyInstaller/utils/hooks/gi.py#L22-L38
but here, serialization to string and subsequent `split()` call hide the fact that we have always been dealing with a single library name, not a list.

rokm committed:
hookutils: gi: fix another regression in get_gi_libdir

GIRepository.Repository.get_shared_library() returns a single string
containing comma-separated library names, so we need to split it
before trying to iterate over it. Also add proper handling of None
case (no shared library associated with the module).

rokm commented:
Now that I properly read the docs I linked, `repo.get_shared_library(module) ` actually returns a "comma-separated list of paths to shared libraries, or None if none are associated". So all we're missing is the `split(',')` call (and handling of `None`)...

danyeaw commented:
Thanks @rokm and @bwoodsend for the team effort 👍
